R/compiler.R

```diff
@@ -319,3 +343,57 @@ substrait_compiler.SubstraitCompiler <- function(object, ...) {
 substrait_compiler.default <- function(object, ...) {
   SubstraitCompiler$new(object, ...)
 }
+
+substrait_call <- function(.fun, ..., .output_type = NULL) {
+  args <- rlang::list2(...)
+  compiler <- current_compiler()
+  template <- substrait$Expression$ScalarFunction$create()
+  compiler$resolve_function(.fun, args, template, .output_type)
+}
+
+substrait_call_agg <- function(.fun, ..., .output_type = NULL) {
+  args <- rlang::list2(...)
+  compiler <- current_compiler()
+  template <- substrait$AggregateFunction$create()
+  compiler$resolve_function(.fun, args, template, .output_type)
+}
+
+substrait_eval <- function(expr) {
+  substrait_eval_quo(rlang::enquo(expr), .data = FALSE)
+}
+
+substrait_eval_data <- function(expr) {
+  substrait_eval_quo(rlang::enquo(expr), .data = TRUE)
+}
+
+substrait_eval_quo <- function(expr_quo, .data = TRUE) {
+  compiler <- current_compiler()
+  rlang::eval_tidy(
+    expr_quo,
+    data = if (is.null(compiler)) NULL else compiler$eval_mask(.data)
+  )
+}
+
+current_compiler <- function() {
+  compiler_context_env$compiler
+}
+
+with_compiler <- function(compiler, expr) {
+  prev_compiler <- current_compiler()
+  on.exit(compiler_context_env$compiler <- prev_compiler)
+  compiler_context_env$compiler <- substrait_compiler(compiler)
+  force(expr)
+}
+
+local_compiler <- function(compiler, .local_envir = parent.frame()) {
+  prev_compiler <- current_compiler()
+  cleanup <- function() {
+    compiler_context_env$compiler <- prev_compiler
+  }
+  cleanup_call <- rlang::call2(cleanup)
+  do.call(base::on.exit, list(cleanup_call, TRUE), envir = .local_envir)
+  compiler_context_env$compiler <- substrait_compiler(compiler)
+}
```

Comment:
Please can you add brief descriptions to these functions in a comment, so it's easier to glance at these functions and know what they do?

Reply:
Thank you, the new documentation you added is great!